ral-facilities/datagateway-api
# Refactor config to be aware of the backend in use

This PR will close #210 

## Description
This is a refactoring for the API's configuration. The root issue described in #210 was caused by calls to config getters in `constants.py` - as this file was imported on startup, there was a requirement for `DB_URL` and `ICAT_URL` to be present. 79d00e8b61509f5dc92b008057d9b3bb91ead3b2 shows the fix for this root issue, with the commit message giving some further explanation. 

3c40ccb546f0434d82366d9a521ac625e6020220 & 9ad2f574ff461e66a59ad0ea2b072d71bf78533b - I've replaced each of the getter functions with a single, generic getter function, taking inspiration from SciGateway Auth on this. Unlike Auth, the config is not hot reloaded as there's no requirement for the API to have this functionality.

Beforehand, the config options included `ICAT_URL` and `DB_URL` - I've changed these to lowercase snake_case to match the formatting of the others. c1a02e5e0aac93906bf4589a17b51657d59c7163 for that change. 

cab681b5e663bcd214aa4f386c45928f8e9a5390 - a function has been added to check that all relevant config options are present in `config.json` at startup. This looks at the backend set and determines which of the backend specific config options must be present for all API functionality to work.

171dbdfee702e8760bd4e130c0738a29675b8618 & 76bcb7d783b8c8f52978a044863e764e63111b2b - I took some further inspiration from SciGateway Auth and implemented an enum class for the API's config options - as the docstring indicates, this is more for ease of use for development rather than bringing any real functional improvements.

## Testing Instructions
I've changed `test_config.py` to reflect these changes, but for manual testing, you should try launching the API with all the config options which startup the API with no issues. Then try removing config option(s) that aren't backend specific, the API should exit when it finds the missing config option, outputting what's missing in the terminal. Do the same for each backend (and their backend specific options), but first set the backend to whichever one you're removing config options from.

- [x] Test that removing config options are spotted at startup, for non-backend specific, ICAT specific and DB specific

## Agile Board Tracking
Connect to #210
